### PR TITLE
First commit for DonutTracker app, a sample for the MAD Skills

### PR DESCRIPTION
**Pull request by chethaase**
Navigation series

This app consists of a list of items (donuts) in a RecyclerView
and a dialog used to enter or edit information about those items.
The sample is mainly useful in conjunction with the MAD Skills series
of articles and videos that use the sample; there are other samples
in this repo that are intended for more general-purpose exmaples
of the Navigation Component.

### CODE REVIEW

File: MADSkillsNavigationSample/app/build.gradle [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    packagingOptions {
+        exclude 'META-INF/atomicfu.kotlin_module'
+    }
+}
+
+configurations {
+    ktlint
+}
+
+dependencies {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/build.gradle:69]: most of these dependencies are not using the latest stable version. please update them

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+class DonutEntryDialogFragment : BottomSheetDialogFragment() {
+
+    private lateinit var donutEntryViewModel: DonutEntryViewModel
+
+    private enum class EditingState {
+        NEW_DONUT,
+        EXISTING_DONUT
+    }
+
+    override fun onActivityCreated(savedInstanceState: Bundle?) {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt:44]: unused function

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        val args: DonutEntryDialogFragmentArgs by navArgs()
+        val editingState =
+            if (args.itemId > 0) EditingState.EXISTING_DONUT
+            else EditingState.NEW_DONUT
+
+        // If we arrived here with an itemId of >= 0, then we are editing an existing item
+        if (editingState == EditingState.EXISTING_DONUT) {
+            // Request to edit an existing item, whose id was passed in as an argument.
+            // Retrieve that item and populate the UI with its details
+            donutEntryViewModel.get(args.itemId).observe(
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt:67]: Use "androidx.lifecycle.observe" instead here. the call becomes:

```
            donutEntryViewModel.get(args.itemId).observe(viewLifecycleOwner) { donutItem ->
                binding.name.setText(donutItem.name)
                binding.description.setText(donutItem.description)
                binding.ratingBar.rating = donutItem.rating.toFloat()
                donut = donutItem
            }
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        binding.doneButton.setOnClickListener {
+            // Grab these now since the Fragment may go away before the setupNotification
+            // lambda below is called
+            val context = requireContext().applicationContext
+            val navController = findNavController()
+
+            donutEntryViewModel.addData(
+                donut?.id ?: 0,
+                binding.name.text.toString(),
+                binding.description.text.toString(), binding.ratingBar.rating.toInt()
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryDialogFragment.kt:89]:
```suggestion
                binding.description.text.toString(), 
                binding.ratingBar.rating.toInt()
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.liveData
+import androidx.lifecycle.viewModelScope
+import com.android.samples.donuttracker.storage.DonutDao
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+class DonutEntryViewModel(val donutDao: DonutDao) : ViewModel() {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutEntryViewModel.kt:27]:
```suggestion
class DonutEntryViewModel(private val donutDao: DonutDao) : ViewModel() {
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutList.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+import kotlinx.android.synthetic.main.donut_list.*
+
+/**
+ * Fragment containing the RecyclerView which shows the current list of donuts being tracked.
+ */
+class DonutList : Fragment() {
+
+    private lateinit var donutListViewModel: DonutListViewModel
+
+    private val adapter = DonutListAdapter(
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutList.kt:39]: use named parameters to increase the readability of the code and change the order of the comment and of the parameter:

```
 private val adapter = DonutListAdapter(
        onEdit = { donut ->
            // Edit selected donut
            findNavController().navigate(
                DonutListDirections.actionDonutListToDonutEntryDialogFragment(donut.id)
            )
        },
        onDelete = {donut ->
            // Delete selected donut
            NotificationManagerCompat.from(requireContext()).cancel(donut.id.toInt())
            donutListViewModel.delete(donut)
        }
    )
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import androidx.recyclerview.widget.DiffUtil
+import androidx.recyclerview.widget.ListAdapter
+import androidx.recyclerview.widget.RecyclerView
+import com.android.samples.donuttracker.databinding.DonutItemBinding
+
+/**
+ * The adapter used by the RecyclerView to display the current list of donuts
+ */
+class DonutListAdapter(private var onEdit: (Donut) -> Unit, private var onDelete: (Donut) -> Unit) :
+    ListAdapter<Donut, DonutListAdapter.DonutListViewHolder>(DonutDiffCallback()) {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt:29]: add import for `DonutListViewHolder`

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    class DonutListViewHolder(
+        val binding: DonutItemBinding,
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt:32]:
```suggestion
        private val binding: DonutItemBinding,
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                onDelete(donut)
+            }
+            binding.root.setOnClickListener {
+                onEdit(donut)
+            }
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DonutListViewHolder {
+        val viewHolder = DonutListViewHolder(
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt:60]: directly return the new object, without creating the val

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+        return viewHolder
+    }
+
+    override fun onBindViewHolder(holder: DonutListViewHolder, position: Int) {
+        holder.bind(getItem(position))
+    }
+}
+
+class DonutDiffCallback : DiffUtil.ItemCallback<Donut>() {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListAdapter.kt:74]: this class can be a private class in `DonutListAdapter`

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.android.samples.donuttracker.storage.DonutDao
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+/**
+ * This ViewModel is used to access the underlying data and to observe changes to it.
+ */
+class DonutListViewModel(val donutDao: DonutDao) : ViewModel() {
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListViewModel.kt:28]:
```suggestion
class DonutListViewModel(private val donutDao: DonutDao) : ViewModel() {
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+    // Users of this ViewModel will observe changes to its donuts list to know when
+    // to redisplay those changes
+    val donuts: LiveData<List<Donut>>
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutListViewModel.kt:32]: inline the initialisation. you don't need to put it in the `init` block

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/Notifier.kt [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+
+/**
+ * Utility class for posting notifications.
+ * This class creates the notification channel (as necessary) and posts to it when requested.
+ */
+object Notifier {
+
+    private val channelId = "Default"
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/Notifier.kt:34]:
```suggestion
    private const val channelId = "Default"
```

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/storage/DonutDatabase.kt [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import android.content.Context
+import androidx.room.Database
+import androidx.room.Room
+import androidx.room.RoomDatabase
+import com.android.samples.donuttracker.Donut
+
+/**
+ * The underlying database where information about the donuts is stored.
+ */
+@Database(entities = arrayOf(Donut::class), version = 1)
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/storage/DonutDatabase.kt:27]:
```suggestion
@Database(entities = [Donut::class], version = 1)
```

File: MADSkillsNavigationSample/app/src/main/res/layout/donut_entry_dialog.xml [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        app:layout_constraintStart_toStartOf="@+id/guideline"
+        app:layout_constraintTop_toTopOf="@+id/guideline5" />
+
+    <EditText
+        android:id="@+id/name"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:ems="10"
+        android:inputType="textPersonName"
+        android:text="Name"
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/res/layout/donut_entry_dialog.xml:55]: Avoid hardcoded text. Move it to resources
If you're using this just for preview use `tools:text`

File: MADSkillsNavigationSample/app/src/main/res/layout/donut_entry_dialog.xml [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        android:inputType="textMultiLine"
+        app:layout_constraintEnd_toStartOf="@+id/guideline4"
+        app:layout_constraintStart_toStartOf="@+id/guideline2"
+        app:layout_constraintTop_toBottomOf="@+id/name" />
+
+    <Button
+        android:id="@+id/doneButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="Done"
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/res/layout/donut_entry_dialog.xml:97]: Avoid hardcoded text. Move it to resources

File: MADSkillsNavigationSample/app/src/main/res/layout/donut_item.xml [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:srcCompat="@android:drawable/presence_online" />
+
+    <TextView
+        android:id="@+id/name"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_margin="8dp"
+        android:text="Name"
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/res/layout/donut_item.xml:38]: Avoid hardcoded text. Move it to resources
If you're using this just for preview use tools:text

File: MADSkillsNavigationSample/app/src/main/res/layout/donut_item.xml [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        app:layout_constraintHorizontal_bias="0.0"
+        app:layout_constraintStart_toEndOf="@+id/thumbnail"
+        app:layout_constraintTop_toBottomOf="@+id/name" />
+
+    <TextView
+        android:id="@+id/rating"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_margin="8dp"
+        android:text="Rating"
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/res/layout/donut_item.xml:62]: Avoid hardcoded text. Move it to resources
If you're using this just for preview use tools:text

File: MADSkillsNavigationSample/app/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<resources>
+    <string name="app_name">Donut Tracker</string>
+    <string name="action_settings">Settings</string>
+    <!-- Strings used for fragments for navigation -->
+    <string name="first_fragment_label">First Fragment</string>
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/res/values/strings.xml:21]: remove all unused strings

File: MADSkillsNavigationSample/build.gradle [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        
+    }
+}
+
+task clean(type: Delete) {
+    delete rootProject.buildDir
+}
+
+ext {
+    roomVersion = '2.2.5'
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/build.gradle:50]: some of these are outdated

File: MADSkillsNavigationSample/build.gradle [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+
+buildscript {
+    ext.kotlin_version = '1.3.71'
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/build.gradle:20]: use kotlin `1.4.10`

**Review by florina-muntenescu — CHANGES_REQUESTED** [review]
mostly nits
run and address lint issues

File: MADSkillsNavigationSample/app/build.gradle (hunk 1 above)
Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:69]: I'm having a hard time figuring out how to determine the latest version for anything that's not in androidx directly (MDC, kotlin, gradle, etc). I'll fix the ones I can figure out.

Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:69]: actually, doing a lint inspection helped immensely, so I was able to update a bunch of the dependencies from that

**Comment by chethaase** [comment]
Hey - I made the changes from the review, mostly fixed through running lint, as you said

File: MADSkillsNavigationSample/app/build.gradle [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+    implementation fileTree(dir: 'libs', include: ['*.jar'])
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'androidx.core:core-ktx:1.3.2'
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/build.gradle:73]: some versions are declared here, some in the other build.gradle file. Be consistent.

File: MADSkillsNavigationSample/app/build.gradle [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+    implementation 'com.google.android.material:material:1.2.1'
+    implementation 'androidx.constraintlayout:constraintlayout:2.0.2'
+    implementation 'androidx.navigation:navigation-fragment-ktx:2.3.0'
+    implementation 'androidx.navigation:navigation-ui-ktx:2.3.0'
+    testImplementation 'junit:junit:4.12'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+
+    implementation 'androidx.fragment:fragment-ktx:1.2.5'
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/build.gradle:82]: i'd suggest keeping the dependencies needed for the app together, followed by the ones needed for tests. but as you're not adding any tests maybe they can just be removed

File: MADSkillsNavigationSample/app/build.gradle [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+        jvmTarget = "1.8"
+    }
+
+    packagingOptions {
+        exclude 'META-INF/atomicfu.kotlin_module'
+    }
+}
+
+configurations {
+    ktlint
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/build.gradle:66]: make sure you run `./gradlew ktlintFormat`

File: MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutList.kt [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+        onEdit = { donut ->
+            findNavController().navigate(
+                DonutListDirections.actionDonutListToDonutEntryDialogFragment(donut.id)
+            )
+        },
+        onDelete = {
+            donut ->
```
Comment (florina-muntenescu) [inline MADSkillsNavigationSample/app/src/main/java/com/android/samples/donuttracker/DonutList.kt:46]: move this donut on the row above

**Review by florina-muntenescu — APPROVED** [review]
Small nits. feel free to merge once addressed

File: MADSkillsNavigationSample/app/build.gradle (hunk 22 above)
Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:73]: Ugh, yeah. Laziness on my part. Couldn't figure out the magic sauce to enter a version number somewhere and refer to it somewhere else (maybe in a different build.gradle file). It's all black magic to me, I find the gradle (especially the kotlin DSL gradle) syntax very mysterious.
but I'll figure it ou

File: MADSkillsNavigationSample/app/build.gradle (hunk 23 above)
Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:82]: good point, will remove

File: MADSkillsNavigationSample/app/build.gradle (hunk 24 above)
Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:66]: that's new to me. I ran gradlew ktlint, didn't know about this one

Comment (chethaase) [reply MADSkillsNavigationSample/app/build.gradle:66]: done, no complaints. Note that I ran (yesterday) lint checks in the IDE (Analyze -> Inspect Code) and fixed various formatting issues that it surfaced.

**Pull request merged**